src/porepy/models/solution_strategy.py

```diff
@@ -724,3 +725,198 @@ def update_time_dependent_ad_arrays(self) -> None:
+        # The time increment of the tangential displacement jump
+        u_t_increment: pp.ad.Operator = pp.ad.time_increment(u_t)
+        zeros_frac = pp.ad.DenseArray(np.zeros(num_cells))
+
+        f_max = pp.ad.Function(pp.ad.maximum, "max_function")
+        f_norm = pp.ad.Function(partial(pp.ad.l2_norm, self.nd - 1), "norm_function")
+        # Heaviside function. The 0 as the second argument to partial() implies
+        # f_heaviside(0)=0, a choice that is not expected to affect the result in this
+        # context.
+        f_heaviside = pp.ad.Function(partial(pp.ad.heaviside, 0), "heaviside_function")
+
+        c_num_as_scalar = self.contact_mechanics_numerical_constant(subdomains)
+
+        c_num = pp.ad.sum_operator_list(
+            [e_i * c_num_as_scalar * e_i.T for e_i in tangential_basis]
+        )
+        tangential_sum = t_t + c_num @ u_t_increment
+
+        max_arg_1 = f_norm(tangential_sum)
+        max_arg_1.set_name("norm_tangential")
+
+        max_arg_2 = f_max(self.friction_bound(subdomains), zeros_frac)
+        max_arg_2.set_name("b_p")
+
+        h_oi = f_heaviside(self.opening_indicator(subdomains))
+        ind = max_arg_1 - max_arg_2
+
+        if self.params.get("adaptive_indicator_scaling", False):
+            # Base on all fracture subdomains
+            all_subdomains = self.mdg.subdomains(dim=self.nd - 1)
+            scale_op = self.contact_traction_estimate(all_subdomains)
+            scale = self.compute_traction_norm(scale_op.value(self.equation_system))
+            ind = ind / pp.ad.Scalar(scale)
+        return ind * h_oi
+
+    def contact_traction_estimate(self, subdomains):
+        """Estimate the characteristic size of contact traction.
+
+        The characteristic fracture traction is estimated as the maximum of the
+        contact traction over the fracture subdomains.
+
+        Parameters:
+            subdomains: List of subdomains where the contact traction is defined.
+
+        Returns:
+            Characteristic fracture traction estimate.
+
+        """
+        # The normal component of the contact traction and the displacement jump
+        t: pp.ad.Operator = self.contact_traction(subdomains)
+        e_n = self.e_i(subdomains, dim=self.nd, i=self.nd - 1)
+
+        u = self.displacement_jump(subdomains) - e_n @ self.fracture_gap(subdomains)
+        c_num = self.contact_mechanics_numerical_constant(subdomains)
+        f_norm = pp.ad.Function(partial(pp.ad.l2_norm, self.nd), "norm_function")
+        return f_norm(t) + f_norm(c_num * u)
+
+    def compute_traction_norm(self, val: np.ndarray) -> float:
+        """Compute a norm of the traction estimate from the vector-valued traction.
+
+        The scalar traction is computed as the p norm of the traction vector.
+
+        Parameters:
+            val: Vector-valued traction.
+
+        Returns:
+            Scalar traction.
+
+        """
+        val = val.clip(1e-8, 1e8)
```

Comment:
I was rather wondering about moderate negative values.

Reply:
But that won't happen, since this method is used on the cell-wise norms. I.e., it takes cell-wise norms and computes one, global value to be used for normalisation.

Reply:
Ah, yes. Now I see how the methods feed each other. OK. I expected something "truely" a traction, but it is this estimate that enters here. All good.